# Activity: Explore hypothesis testing

## Introduction

You work for an environmental think tank called Repair Our Air (ROA). ROA is formulating policy recommendations to improve the air quality in America, using the Environmental Protection Agency's Air Quality Index (AQI) to guide their decision making. An AQI value close to 0 signals "little to no" public health concern, while higher values are associated with increased risk to public health. 

They've tasked you with leveraging AQI data to help them prioritize their strategy for improving air quality in America.

ROA is considering the following decisions. For each, construct a hypothesis test and an accompanying visualization, using your results of that test to make a recommendation:

1. ROA is considering a metropolitan-focused approach. Within California, they want to know if the mean AQI in Los Angeles County is statistically different from the rest of California.
2. With limited resources, ROA has to choose between New York and Ohio for their next regional office. Does New York have a lower AQI than Ohio?
3. A new policy will affect those states with a mean AQI of 10 or greater. Will Michigan be affected by this new policy?

**Notes:**
1. For your analysis, you'll default to a 5% level of significance.
2. Throughout the lab, for two-sample t-tests, use Welch's t-test (i.e., setting the `equal_var` parameter to `False` in `scipy.stats.ttest_ind()`). This will account for the possibly unequal variances between the two groups in the comparison.

## Step 1: Imports

To proceed with your analysis, import `pandas` and `numpy`. To conduct your hypothesis testing, import `stats` from `scipy`.

#### Import Packages

In [1]:
# Import relevant packages

### YOUR CODE HERE ###
import pandas as pd
import numpy as np
from scipy import stats

You are also provided with a dataset with national Air Quality Index (AQI) measurements by state over time for this analysis. `Pandas` was used to import the file `c4_epa_air_quality.csv` as a dataframe named `aqi`. As shown in this cell, the dataset has been automatically loaded in for you. You do not need to download the .csv file, or provide more code, in order to access the dataset and proceed with this lab. Please continue with this activity by completing the following instructions.

**Note:** For purposes of your analysis, you can assume this data is randomly sampled from a larger population.

#### Load Dataset

In [2]:
# RUN THIS CELL TO IMPORT YOUR DATA.

### YOUR CODE HERE ###
aqi = pd.read_csv('c4_epa_air_quality.csv')

## Step 2: Data Exploration

### Before proceeding to your deliverables, explore your datasets.

Use the following space to surface descriptive statistics about your data. In particular, explore whether you believe the research questions you were given are readily answerable with this data.

In [3]:
# Explore your dataframe `aqi` here:

### YOUR CODE HERE ###
aqi.describe()

Unnamed: 0  arithmetic_mean         aqi
count  260.000000       260.000000  260.000000
mean   129.500000         0.403169    6.757692
std     75.199734         0.317902    7.061707
min      0.000000         0.000000    0.000000
25%     64.750000         0.200000    2.000000
50%    129.500000         0.276315    5.000000
75%    194.250000         0.516009    9.000000
max    259.000000         1.921053   50.000000

In [5]:
aqi.head(10)

Unnamed: 0  date_local    state_name   county_name      city_name  \
0           0  2018-01-01       Arizona      Maricopa        Buckeye   
1           1  2018-01-01          Ohio       Belmont      Shadyside   
2           2  2018-01-01       Wyoming         Teton  Not in a city   
3           3  2018-01-01  Pennsylvania  Philadelphia   Philadelphia   
4           4  2018-01-01          Iowa          Polk     Des Moines   
5           5  2018-01-01        Hawaii      Honolulu  Not in a city   
6           6  2018-01-01        Hawaii      Honolulu  Not in a city   
7           7  2018-01-01  Pennsylvania          Erie           Erie   
8           8  2018-01-01        Hawaii      Honolulu       Honolulu   
9           9  2018-01-01      Colorado       Larimer   Fort Collins   

                                     local_site_name   parameter_name  \
0                                            BUCKEYE  Carbon monoxide   
1                                          Shadyside  Carbon monoxide   
2  Yellowstone National Park - Old Faithful Snow ...  Carbon monoxide   
3                             North East Waste (NEW)  Carbon monoxide   
4                                          CARPENTER  Carbon monoxide   
5                                            Kapolei  Carbon monoxide   
6                                            Kapolei  Carbon monoxide   
7                                                NaN  Carbon monoxide   
8                                           Honolulu  Carbon monoxide   
9                      Fort Collins - CSU - S. Mason  Carbon monoxide   

    units_of_measure  arithmetic_mean  aqi  
0  Parts per million         0.473684    7  
1  Parts per million         0.263158    5  
2  Parts per million         0.111111    2  
3  Parts per million         0.300000    3  
4  Parts per million         0.215789    3  
5  Parts per million         0.994737   14  
6  Parts per million         0.200000    2  
7  Parts per million         0.200000    2  
8  Parts per million         0.400000    5  
9  Parts per million         0.300000    6

In [6]:
aqi.shape

(260, 10)

<details>
  <summary><h4><strong>HINT 1</strong></h4></summary>

  Consider referring to the material on descriptive statisics.
</details>

<details>
  <summary><h4><strong>HINT 2</strong></h4></summary>

  Consider using `pandas` or `numpy` to explore the `aqi` dataframe.
</details>

<details>
  <summary><h4><strong>HINT 3</strong></h4></summary>

Any of the following functions may be useful:
- `pandas`: `describe()`,`value_counts()`,`shape()`, `head()`
- `numpy`: `unique()`,`mean()`
    
</details>

#### **Question 1: From the preceding data exploration, what do you recognize?**

A: There are a total of 260 unique data readings in the AQI dataset, and there are no missing values present at all.

## Step 3. Statistical Tests

Before you proceed, recall the following steps for conducting hypothesis testing:

1. Formulate the null hypothesis and the alternative hypothesis.<br>
2. Set the significance level.<br>
3. Determine the appropriate test procedure.<br>
4. Compute the p-value.<br>
5. Draw your conclusion.

### Hypothesis 1: ROA is considering a metropolitan-focused approach. Within California, they want to know if the mean AQI in Los Angeles County is statistically different from the rest of California.

Before proceeding with your analysis, it will be helpful to subset the data for your comparison.

In [8]:
# Create dataframes for each sample being compared in your test

### YOUR CODE HERE ###
first_data = aqi[aqi['county_name'] == "Los Angeles"]
second_data = aqi[(aqi['state_name'] == "California") & (aqi['county_name'] != "Los Angeles")]

In [9]:
first_data.head()

Unnamed: 0  date_local  state_name  county_name      city_name  \
33           33  2018-01-01  California  Los Angeles      Lancaster   
42           42  2018-01-01  California  Los Angeles  Santa Clarita   
61           61  2018-01-01  California  Los Angeles       Pasadena   
76           76  2018-01-01  California  Los Angeles    Los Angeles   
109         109  2018-01-01  California  Los Angeles    Los Angeles   

                   local_site_name   parameter_name   units_of_measure  \
33       Lancaster-Division Street  Carbon monoxide  Parts per million   
42                   Santa Clarita  Carbon monoxide  Parts per million   
61                        Pasadena  Carbon monoxide  Parts per million   
76                    LAX Hastings  Carbon monoxide  Parts per million   
109  Los Angeles-North Main Street  Carbon monoxide  Parts per million   

     arithmetic_mean  aqi  
33          0.394737    7  
42          0.394737    7  
61          0.789474   16  
76          0.863158   17  
109         0.994737   17

In [10]:
second_data.head()

Unnamed: 0  date_local  state_name     county_name      city_name  \
16          16  2018-01-01  California  San Bernardino        Ontario   
18          18  2018-01-01  California      Sacramento   Arden-Arcade   
26          26  2018-01-01  California          Orange       La Habra   
27          27  2018-01-01  California         Alameda  Not in a city   
34          34  2018-01-01  California          Fresno         Fresno   

                 local_site_name   parameter_name   units_of_measure  \
16  Ontario Near Road (Etiwanda)  Carbon monoxide  Parts per million   
18     Sacramento-Del Paso Manor  Carbon monoxide  Parts per million   
26                      La Habra  Carbon monoxide  Parts per million   
27        Berkeley- Aquatic Park  Carbon monoxide  Parts per million   
34              Fresno - Garland  Carbon monoxide  Parts per million   

    arithmetic_mean  aqi  
16         0.747368   11  
18         0.752632   16  
26         0.673684   13  
27         1.088889   15  
34         1.000000   15

<details>
  <summary><h4><strong>HINT 1</strong></h4></summary>

  Consider referencing the material on subsetting dataframes.  
</details>

<details>
  <summary><h4><strong>HINT 2</strong></h4></summary>

  Consider creating two dataframes, one for Los Angeles, and one for all other California observations.
</details>

<details>
  <summary><h4><strong>HINT 3</strong></h4></summary>

For your first dataframe, filter to `county_name` of `Los Angeles`. For your second dataframe, filter to `state_name` of `Calfornia` and `county_name` not equal to `Los Angeles`.
    
</details>

#### Formulate your hypothesis:

**Formulate your null and alternative hypotheses:**

*   $H_0$: There is no difference in the mean AQI between Los Angeles County and the rest of California.
*   $H_A$: There is a difference in the mean AQI between Los Angeles County and the rest of California.


#### Set the significance level:

In [11]:
# For this analysis, the significance level is 5%

### YOUR CODE HERE
sig_level = 0.05

#### Determine the appropriate test procedure:

Here, you are comparing the sample means between two independent samples. Therefore, you will utilize a **two-sample  𝑡-test**.

#### Compute the P-value

In [12]:
# Compute your p-value here

### YOUR CODE HERE ###
p_val = stats.ttest_ind(a=first_data['aqi'], b=second_data['aqi'], equal_var=False)
p_val

Ttest_indResult(statistic=2.1107010796372014, pvalue=0.049839056842410995)

<details>
  <summary><h4><strong>HINT 1</strong></h4></summary>

  Consider referencing the material on how to perform a two-sample t-test.
</details>

<details>
  <summary><h4><strong>HINT 2</strong></h4></summary>

  In `ttest_ind()`, a is the aqi column from our "Los Angeles" dataframe, and b is the aqi column from the "Other California" dataframe.
</details>

<details>
  <summary><h4><strong>HINT 3</strong></h4></summary>

  Be sure to set `equal_var` = False.

</details>

#### **Question 2. What is your P-value for hypothesis 1, and what does this indicate for your null hypothesis?**

A: The P-value for hypothesis 1 is 0.0498, and because it is less than the significance level of 5%, reject the null hypothesis, given that there is a difference in the mean AQI for Los Angeles County and the rest of California.

### Hypothesis 2: With limited resources, ROA has to choose between New York and Ohio for their next regional office. Does New York have a lower AQI than Ohio?

Before proceeding with your analysis, it will be helpful to subset the data for your comparison.

In [13]:
# Create dataframes for each sample being compared in your test

### YOUR CODE HERE ###
first_data = aqi[aqi['state_name'] == "New York"]
second_data = aqi[aqi['state_name'] == "Ohio"]

In [14]:
first_data.head()

Unnamed: 0  date_local state_name county_name    city_name  \
90           90  2018-01-01   New York        Erie  Cheektowaga   
113         113  2018-01-01   New York       Bronx     New York   
124         124  2018-01-01   New York      Monroe    Rochester   
167         167  2018-01-01   New York    New York     New York   
173         173  2018-01-01   New York      Queens     New York   

              local_site_name   parameter_name   units_of_measure  \
90          Buffalo Near-Road  Carbon monoxide  Parts per million   
113           PFIZER LAB SITE  Carbon monoxide  Parts per million   
124               ROCHESTER 2  Carbon monoxide  Parts per million   
167                      CCNY  Carbon monoxide  Parts per million   
173  Queens College Near Road  Carbon monoxide  Parts per million   

     arithmetic_mean  aqi  
90          0.252632    3  
113         0.289474    3  
124         0.200000    2  
167         0.200000    2  
173         0.273684    3

In [15]:
second_data.head()

Unnamed: 0  date_local state_name county_name   city_name local_site_name  \
1            1  2018-01-01       Ohio     Belmont   Shadyside       Shadyside   
12          12  2018-01-01       Ohio    Hamilton  Cincinnati      Taft NCore   
22          22  2018-01-01       Ohio       Stark      Canton          Canton   
51          51  2018-01-01       Ohio      Summit       Akron    NIHF STEM MS   
59          59  2018-01-01       Ohio    Cuyahoga   Cleveland  GT Craig NCore   

     parameter_name   units_of_measure  arithmetic_mean  aqi  
1   Carbon monoxide  Parts per million         0.263158    5  
12  Carbon monoxide  Parts per million         0.252632    3  
22  Carbon monoxide  Parts per million         0.394737    6  
51  Carbon monoxide  Parts per million         0.083333    3  
59  Carbon monoxide  Parts per million         0.250000    3

<details>
  <summary><h4><strong>HINT 1</strong></h4></summary>

  Consider referencing the materials on subsetting dataframes.  
</details>

<details>
  <summary><h4><strong>HINT 2</strong></h4></summary>

  Consider creating two dataframes, one for New York, and one for Ohio observations.
</details>

<details>
  <summary><h4><strong>HINT 3</strong></h4></summary>

For your first dataframe, filter to `state_name` of `New York`. For your second dataframe, filter to `state_name` of `Ohio`.
    
</details>

#### Formulate your hypothesis:

**Formulate your null and alternative hypotheses:**

*   $H_0$: The mean AQI of New York is greater than or equal to that of Ohio.
*   $H_A$: The mean AQI of New York is **below** that of Ohio.


#### Significance Level (remains at 5%)

#### Determine the appropriate test procedure:

Here, you are comparing the sample means between two independent samples in one direction. Therefore, you will utilize a **two-sample  𝑡-test**.

#### Compute the P-value

In [16]:
# Compute your p-value here

### YOUR CODE HERE ###
p_val = stats.ttest_ind(a=first_data['aqi'], b=second_data['aqi'], equal_var=False, alternative='less')
p_val

Ttest_indResult(statistic=-2.025951038880333, pvalue=0.030446502691934697)

<details>
  <summary><h4><strong>HINT 1</strong></h4></summary>

  Consider referencing the material on how to perform a two-sample t-test.
</details>

<details>
  <summary><h4><strong>HINT 2</strong></h4></summary>

  In `ttest_ind()`, a is the aqi column from the "New York" dataframe, an b is the aqi column from the "Ohio" dataframe.
</details>

<details>
  <summary><h4><strong>HINT 3</strong></h4></summary>

  You can assign `tstat`, `pvalue` to the output of `ttest_ind`. Be sure to include `alternative = less` as part of your code.  

</details>

#### **Question 3. What is your P-value for hypothesis 2, and what does this indicate for your null hypothesis?**

A: The P-value for hypothesis is 0.0304, and because it is less than the significance level of 5%, reject the null hypothesis, given that the mean AQI of New York is below the mean AQI of Ohio.

###  Hypothesis 3: A new policy will affect those states with a mean AQI of 10 or greater. Will Michigan be affected by this new policy?

Before proceeding with your analysis, it will be helpful to subset the data for your comparison.

In [17]:
# Create dataframes for each sample being compared in your test

### YOUR CODE HERE ###
subdata = aqi[aqi['state_name'] == "Michigan"]
subdata.head()

Unnamed: 0  date_local state_name county_name   city_name  \
65           65  2018-01-01   Michigan       Wayne     Livonia   
122         122  2018-01-01   Michigan       Wayne     Detroit   
123         123  2018-01-01   Michigan       Wayne     Detroit   
129         129  2018-01-01   Michigan       Wayne     Detroit   
192         192  2018-01-01   Michigan       Wayne  Allen Park   

              local_site_name   parameter_name   units_of_measure  \
65                 LIVONIA-NR  Carbon monoxide  Parts per million   
122               West corner  Carbon monoxide  Parts per million   
123  MARK TWAIN MIDDLE SCHOOL  Carbon monoxide  Parts per million   
129                  ELIZA-NR  Carbon monoxide  Parts per million   
192                Allen Park  Carbon monoxide  Parts per million   

     arithmetic_mean  aqi  
65          0.338889    5  
122         0.394737    8  
123         0.515789    9  
129         0.616667   11  
192         0.811111   13

<details>
  <summary><h4><strong>HINT 1</strong></h4></summary>

  Consider referencing the material on subsetting dataframes.  
</details>

<details>
  <summary><h4><strong>HINT 2</strong></h4></summary>

  Consider creating one dataframe which only includes Michigan.
</details>

#### Formulate your hypothesis:

**Formulate your null and alternative hypotheses here:**

*   $H_0$: The mean AQI of Michigan is less than or equal to 10.
*   $H_A$: The mean AQI of Michigan is greater than 10.


#### Significance Level (remains at 5%)

#### Determine the appropriate test procedure:

Here, you are comparing one sample mean relative to a particular value in one direction. Therefore, you will utilize a **one-sample  𝑡-test**. 

#### Compute the P-value

In [21]:
# Compute your p-value here

### YOUR CODE HERE ###
p_val = stats.ttest_1samp(a=subdata['aqi'], popmean=10, alternative='greater')
p_val

Ttest_1sampResult(statistic=-1.7395913343286131, pvalue=0.9399405193140109)

<details>
  <summary><h4><strong>HINT 1</strong></h4></summary>

  Consider referencing the material on how to perform a one-sample t-test.
</details>

<details>
  <summary><h4><strong>HINT 2</strong></h4></summary>

  In `ttest_1samp)`, you are comparing the aqi column from your Michigan data relative to 10, the new policy threshold.
</details>

<details>
  <summary><h4><strong>HINT 3</strong></h4></summary>

  You can assign `tstat`, `pvalue` to the output of `ttest_1samp`. Be sure to include `alternative = greater` as part of your code.  

</details>

#### **Question 4. What is your P-value for hypothesis 3, and what does this indicate for your null hypothesis?**

A: The P-value for hypothesis 3 is 0.9399, and because it is greater than the significance level of 5%, fail to reject the null hypothesis, meaning that the mean AQI of Michigan is not greater than 10.

## Step 4. Results and Evaluation

Now that you've completed your statistical tests, you can consider your hypotheses and the results you gathered.

#### **Question 5. Did your results show that the AQI in Los Angeles County was statistically different from the rest of California?**

A: The results show that the AQI values in Los Angeles County is statistically different from the other counties of California, given that the P-value is less than the significance level of 5%.

#### **Question 6. Did New York or Ohio have a lower AQI?**

A: Since the P-value is less than the significance level of 5%, New York has a lower AQI reading than Ohio.

#### **Question 7: Will Michigan be affected by the new policy impacting states with a mean AQI of 10 or greater?**



A: Because the P-value is greater than the significance level of 5%, Michigan will not be affected by the new policy impacting states with a mean AQI of 10 or higher.

# Conclusion

**What are key takeaways from this lab?**

A: The key takeaways from this lab are that at the 5% significance level, the mean AQI value for Los Angeles county is significantly different from the other counties in California, and the mean AQI for New York is less than the mean AQI for Ohio. However, it does not apply to Michigan having a mean AQI value of 10 or higher.

**What would you consider presenting to your manager as part of your findings?**

A: For each test conducted, state the null and alternative hypothesis, then based upon the P-value obtained, describe the conclusion and state whether or not the null hypothesis can be rejected. Also specify the type of test performed whether it is a one-tailed or two-tailed test performed, and how that is conducted using the appropriate functions from the stats library.

**What would you convey to external stakeholders?**

A: State the standard level of significance and the conclusions drawn for each hypothesis test performed. Providing the sample initial statistics for each experiment can provide some context and background for the stakeholders to better understand the meaning of the results.

**Congratulations!** You've completed this lab. However, you may not notice a green check mark next to this item on Coursera's platform. Please continue your progress regardless of the check mark. Just click on the "save" icon at the top of this notebook to ensure your work has been logged.